cpp/src/arrow/compute/exec/project_benchmark.cc

```diff
@@ -0,0 +1,104 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "benchmark/benchmark.h"
+
+#include "arrow/compute/cast.h"
+#include "arrow/compute/exec.h"
+#include "arrow/compute/exec/expression.h"
+#include "arrow/compute/exec/options.h"
+#include "arrow/compute/exec/test_util.h"
+#include "arrow/dataset/partition.h"
+#include "arrow/testing/future_util.h"
+#include "arrow/testing/generator.h"
+#include "arrow/testing/gtest_util.h"
+#include "arrow/type.h"
+
+namespace arrow {
+namespace compute {
+
+static constexpr int64_t total_batch_size = 1e6;
+
+void SetArgs(benchmark::internal::Benchmark* bench) {
+  for (auto batch_size = 1; batch_size <= 1e6; batch_size *= 10) {
+    auto num_batches = total_batch_size / batch_size;
+    bench->ArgNames({"batch_size, num_batches"})
+        ->Args({batch_size, num_batches})
+        ->UseRealTime();
+  }
+}
+
+/*
+should be able to vary
+1. batch size
+2. expression complexity
+*/
+static void ExecuteScalarProjectionOverhead(benchmark::State& state, Expression expr) {
+  const auto batch_size = static_cast<int32_t>(state.range(0));
+  const auto num_batches = static_cast<int32_t>(state.range(1));
+
+  auto data = MakeRandomBatches(schema({field("i64", int64()), field("bool", boolean())}),
+                                num_batches, batch_size);
+  ExecContext ctx;
+
+  for (auto _ : state) {
+    ASSERT_OK_AND_ASSIGN(auto plan, ExecPlan::Make());
+    // CountOptions options(CountOptions::ONLY_VALID);
+    AsyncGenerator<util::optional<ExecBatch>> sink_gen;
+    ASSERT_OK(Declaration::Sequence(
+                  {
+                      {"source",
+                       SourceNodeOptions{data.schema,
+                                         data.gen(/*parallel=*/false, /*slow=*/false)},
+                       "custom_source_label"},
+                      {"project", ProjectNodeOptions{{
+                                      field_ref("bool"),
```

Comment:
`field_ref(bool)` is here, but may not contribute must to the tests.

Reply:
Let's remove so that the benchmark focuses purely on expr.